src/main/resources/cpp-modern-json/source.mustache

```diff
@@ -22,5 +28,37 @@ namespace {{this}} {
 {{#modelNamespaceDeclarations}}
 }
 {{/modelNamespaceDeclarations}}
+
+{{#hasChildren}}
+namespace nlohmann {
+namespace _models = {{modelNamespace}};
+void adl_serializer<std::shared_ptr<_models::{{classname}}>>::from_json(
+    const json &j, std::shared_ptr<_models::{{classname}}> &ptr)
+{
+    if (j.is_null()) {
+      ptr.reset();
+    } else {
+      const auto base = j.get<_models::{{classname}}>();
+      {{#discriminator}}
+      switch(base.{{propertyBaseName}}) {
+      {{#mappedModels}}
+        case decltype(base.{{propertyBaseName}})::{{mappingName}}:
+          ptr = std::make_shared<_models::{{modelName}}>(j.get<_models::{{modelName}}>());
+          return;
+      {{/mappedModels}}
+      }
+      UNREACHABLE_CODE
+      {{/discriminator}}
+    }
+}
+
+void adl_serializer<std::shared_ptr<_models::{{classname}}>>::to_json(
+    json &j, const std::shared_ptr<_models::{{classname}}> &ptr)
+{
+    // no special handling needed -> redirect to to_json in serialization.h
```

Comment:
Is this needed at all then?

Reply:
Nope. We specialize the struct here. Not the functions. So the serialization will use this exact serializer and won't find a `to_json` function.
Turns out, however, that we need the `to_json` function anyway
(e.g.: We when posting a `SecretsBulkRequest`, which holds a vector of `SecretsBulkRequestItem`, consisting of a `std::shared_ptr<CreateBaseSecretRequest>`, which will need to be serialized to the correct json type. Using the default shared_ptr serialization, this would not include the special fields of the subclasses)